### PR TITLE
[DO NOT MERGE] Sdw master device review v2

### PR DESCRIPTION
**Pull request by bardliao**
This is an alternative PR to #2072 . The first 3 commits are already merged by Vinod. See #2104 .
The PR address all Vinod's commands except master id.

### CODE REVIEW

File: include/linux/soundwire/sdw.h [hunk 1]
```diff
@@ -789,6 +789,7 @@ struct sdw_master_ops {
  * struct sdw_bus - SoundWire bus
  * @dev: Master linux device
  * @link_id: Link id number, can be 0 to N, unique for each Master
+ * @id: Unique id for each bus
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h]: maybe reword:

* @link_id: Link id number, can be 0 to N, unique for each Master in the controller scope.
 * @id: bus system-wide unique id

File: include/linux/soundwire/sdw.h [hunk 2]
```diff
@@ -812,6 +813,7 @@ struct sdw_master_ops {
  */
 struct sdw_bus {
 	struct device *dev;
+	int id;
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h]: make sure the member and kernel doc are in the same order.
This will report errors with make W=1

File: drivers/soundwire/Makefile [hunk 3]
```diff
@@ -4,7 +4,7 @@
 #
 
 #Bus Objs
-soundwire-bus-objs := bus_type.o bus.o slave.o mipi_disco.o stream.o
```
Comment (plbossart) [inline drivers/soundwire/Makefile:7]: comment on commit message: you need to remove the part on 'abusing platform devices'. This is no longer true.

Suggested text:

In the existing SoundWire code, Master Devices are not explicitly
represented - only SoundWire Slave Devices are exposed (the use of
capital letters follows the SoundWire specification conventions).

With the existing code, the bus is handled without using a proper device, and bus->dev typically points to a platform device. The right thing to do as discussed in multiple reviews is use a device for each bus.

The sdw_master_device addition is done with minimal internal plumbing
and not exposed externally. The existing API based on
sdw_bus_master_add() and sdw_bus_master_delete() will deal with the
sdw_master_device life cycle, which minimizes changes to existing
drivers.

Note that the Intel code will be modified in follow-up patches (no
impact on any platform since the connection with ASoC is not supported
upstream so far).

File: drivers/soundwire/bus.c [hunk 4]
```diff
@@ -157,7 +164,8 @@ static int sdw_delete_slave(struct device *dev, void *data)
  */
 void sdw_bus_master_delete(struct sdw_bus *bus)
 {
-	device_for_each_child(bus->dev, NULL, sdw_delete_slave);
+	device_for_each_child(&bus->md->dev, NULL, sdw_delete_slave);
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: you can still use bus->dev here, as long as it's properly initialized.

File: drivers/soundwire/master.c [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		 * On err, don't free but drop ref as this will be freed
+		 * when release method is invoked.
+		 */
+		put_device(&md->dev);
+		goto device_register_err;
+	}
+
+	/* add shortcuts to improve code readability/compactness */
+	md->bus = bus;
+	bus->md = md;
```
Comment (plbossart) [inline drivers/soundwire/master.c]: where is bus->dev set? this is needed in all the rest of the code.

File: drivers/soundwire/slave.c [hunk 6]
```diff
@@ -32,7 +32,7 @@ static int sdw_slave_add(struct sdw_bus *bus,
 
 	/* Initialize data structure */
 	memcpy(&slave->id, id, sizeof(*id));
-	slave->dev.parent = bus->dev;
+	slave->dev.parent = &bus->md->dev;
```
Comment (plbossart) [inline drivers/soundwire/slave.c]: again you don't need to change this if bus->dev is set to &bus->md->dev

File: include/linux/soundwire/sdw.h [hunk 7]
```diff
@@ -787,7 +800,8 @@ struct sdw_master_ops {
 
 /**
  * struct sdw_bus - SoundWire bus
- * @dev: Master linux device
+ * @dev: Soundwire bus linux device
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h]: dev: shortcut to &bus->md->dev to avoid changing the entire code.

File: include/linux/soundwire/sdw.h [hunk 8]
```diff
@@ -632,6 +632,19 @@ struct sdw_slave {
 
 #define dev_to_sdw_dev(_dev) container_of(_dev, struct sdw_slave, dev)
 
+/**
+ * struct sdw_master_device - SoundWire 'Master Device' representation
+ * @dev: Linux device for this Master
+ * @bus: Bus handle shortcut to improve readability (same as container_of)
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h]: it's not the same as container_of, since the sdw_master_device is a pointer, not a structure in struct sdw_bus().
You cannot use container_of to go from md to bus without this link.

File: drivers/soundwire/master.c [hunk 9]
```diff
@@ -14,9 +15,15 @@ static void sdw_master_device_release(struct device *dev)
...
 
+static const struct dev_pm_ops master_dev_pm = {
+	SET_RUNTIME_PM_OPS(pm_generic_runtime_suspend,
+			   pm_generic_runtime_resume, NULL)
+};
+
 struct device_type sdw_master_type = {
 	.name =		"soundwire_master",
 	.release =	sdw_master_device_release,
+	.pm = &master_dev_pm,
```
Comment (plbossart) [inline drivers/soundwire/master.c:26]: commit message:

we need to enable runtime_pm on master device with generic helpers, so that a Slave-initiated wake is propagated to the bus parent.

File: drivers/soundwire/bus.c (hunk 4 above)
Comment (bardliao) [reply drivers/soundwire/bus.c]: That's a major change in this series. Now, md->dev is a child of bus->dev. That is to say bus->dev is no longer &md->dev anymore.
https://lkml.org/lkml/2020/5/11/175

File: drivers/soundwire/master.c (hunk 5 above)
Comment (bardliao) [reply drivers/soundwire/master.c]: It is ``sdw->cdns.bus.dev = &pdev->dev;`` in intel_bus_probe() and ``ctrl->bus.dev = dev`` in qcom_swrm_probe()

File: drivers/soundwire/slave.c (hunk 6 above)
Comment (bardliao) [reply drivers/soundwire/slave.c]: Now md->dev is a child of bus->dev. And when we call ``pm_runtime_get_sync(slave->bus->dev);`` in sdw_nread/nwrite  what we really want is to wake up the intel device not master device, right?

File: include/linux/soundwire/sdw.h (hunk 8 above)
Comment (bardliao) [reply include/linux/soundwire/sdw.h]: Ah, yes. Thanks.

File: include/linux/soundwire/sdw.h (hunk 1 above)
Comment (bardliao) [reply include/linux/soundwire/sdw.h]: updated.

File: include/linux/soundwire/sdw.h (hunk 2 above)
Comment (bardliao) [reply include/linux/soundwire/sdw.h]: Thanks, it was a mistake of rebasing. :)

File: drivers/soundwire/Makefile (hunk 3 above)
Comment (bardliao) [reply drivers/soundwire/Makefile:7]: > With the existing code, the bus is handled without using a proper device, and bus->dev typically points to a platform device. The right thing to do as discussed in multiple reviews is use a device for each bus.

Updated. But I am concerned about this sentence. See https://lkml.org/lkml/2020/5/11/428

File: drivers/soundwire/bus.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: this makes absolutely no sense to me, but I am not going to argue. If this is what it takes to get the code upstream, so be it.

**Comment by plbossart** [comment]
@bardliao please go ahead and deal with upstream direclty, I don't agree with the directions but I will follow the 'disagree-and-commit' principle

File: drivers/soundwire/Makefile (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/Makefile:7]: yes exactly, this is inconsistent with GregKH's advice!

**Comment by bardliao** [comment]
@plbossart I did some change and set bus->dev = &md->dev. Basically, bus->dev is used as parent to create Master device before Master device and point to Master device once it is create. I did the same change on #2088 and will send the patches upstream if the test result of #2088 is good.

**Comment by plbossart** [comment]
> @plbossart I did some change and set bus->dev = &md->dev. Basically, bus->dev is used as parent to create Master device before Master device and point to Master device once it is create. I did the same change on #2088 and will send the patches upstream if the test result of #2088 is good.

@bardliao why would you use two definitions for the bus->dev? We already provide the 'parent' argument, it seems very odd to me to overload the bus->dev field for two different purposes.

**Comment by bardliao** [comment]
> > @plbossart I did some change and set bus->dev = &md->dev. Basically, bus->dev is used as parent to create Master device before Master device and point to Master device once it is create. I did the same change on #2088 and will send the patches upstream if the test result of #2088 is good.
> 
> @bardliao why would you use two definitions for the bus->dev? We already provide the 'parent' argument, it seems very odd to me to overload the bus->dev field for two different purposes.

I removed the 'parent' argument in this version. Let me put it back.

**Comment by bardliao** [comment]
Close it as they are applied

**Comment by bardliao** [comment]
Reopen it as ``soundwire: intel: transition to 3 steps initialization`` is not upstream yet.

**Comment by bardliao** [comment]
@plbossart Does https://github.com/thesofproject/linux/pull/2107/commits/5d1b316a59a0943891a5edc571a7de4b8e0c0751 look good to you?

**Comment by plbossart** [comment]
> @plbossart Does [5d1b316](https://github.com/thesofproject/linux/commit/5d1b316a59a0943891a5edc571a7de4b8e0c0751) look good to you?

Might be better to do an upstream merge and resubmit a new PR, there are too many patches already accepted. Let me do this today, that way we have a clearer picture of what remains to be sent upstream.

**Comment by plbossart** [comment]
> @plbossart Does [5d1b316](https://github.com/thesofproject/linux/commit/5d1b316a59a0943891a5edc571a7de4b8e0c0751) look good to you?

looks mostly good, I added cosmetic/renaming comments but that seems fine with me.

**Comment by plbossart** [comment]
> > @plbossart Does [5d1b316](https://github.com/thesofproject/linux/commit/5d1b316a59a0943891a5edc571a7de4b8e0c0751) look good to you?
> 
> Might be better to do an upstream merge and resubmit a new PR, there are too many patches already accepted. Let me do this today, that way we have a clearer picture of what remains to be sent upstream.

Actually we can't do a rebase without these patches, upstream and topic/sof-dev have different defintions for sdw_master_device so we need to merge this series very very soon, since it'll gate the rebase/merge process...

**Comment by bardliao** [comment]
replaced by #2130
